keps/sig-storage/20190125-online-growing-persistent-volume-size.md

```diff
@@ -0,0 +1,299 @@
+
+### Notes
+
+- Currently we only support offline resizing for `xfs`, `ext3`, `ext4`. Online resizing of `ext3`,`ext4` was introduced in [Linux kernel-3.3](https://www.ibm.com/developerworks/library/l-33linuxkernel/), and `xfs` has always supported growing mounted partitions (in fact, currently there is no way to expand an unmounted `xfs` file system), so they are all safe for online resizing. If a user tries to expand a volume with other formats an error event will be reported for the pod using it.
+
+- This feature is protected by an alpha feature gate `ExpandOnlinePersistentVolumes` in v1.11. We separate this feature gate from the offline resizing gate `ExpandPersistentVolumes`, if a user wants to enable this feature, `ExpandPersistentVolumes` must be enabled first.
+
+### Implementation Details
+
+The key point of online file system resizing is how kubelet discovers which  PVCs need file system resizing. We achieve this goal by reusing the reprocess mechanism of `VolumeManager`'s `DesiredStateOfWorldPopulator`. In detail, kubelet synchronizes pods' status periodically, and during each loop, `DesiredStateOfWorldPopulator` will reprocess each pod's volumes and add them to `ActualStateOfWorld`. If a volume refers to a PVC, it will fetch both PV and PVC objects from apiserver. So we can check online resize request on these latest fetched objects.
+
+#### Online File System Resize in Kubelet
+
+##### Volume Resize Request Detect
+
+As mentioned before,`DesiredStateOfWorldPopulator` loops through all alive pods in `PodManager` and reprocess each pod's volumes periodically. If a volume refers to a `PersistentVolumeClaim`, it will fetch the latest PV and PVC objects from apiserver, see [here](https://github.com/kubernetes/kubernetes/blob/6b64c07bafd13f7a2f4396526f60a189d407e363/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L439) and [here](https://github.com/kubernetes/kubernetes/blob/2f13ffb056488dd34c92a0e0e1c57bef0a45f27d/pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go#L485). The reason for this is we need the PV object to create `VolumeSpec`, and we need the PVC object to know which PV it is bound to.
+
+With these latest objects, we can check whether a PVC requires an online resizing operation or not. A PVC requires online file system resizing if:
+
+* `PVC.Status.Capacity` is less than  `PV.Spec.Capacity`
+* `PVC.Status.Condition` has a `PersistentVolumeClaimFileSystemResizePending` condition
+
+To mark a volume as file system resizing required, we add a `fsResizeRequired` field to `attachedVolume.mountedPod`, and add a `MarkFSResizeRequired` method to `ActualStateOfWorld`. If `DesiredStateOfWorldPopulator` discovers a volume requires file system resizing, it invokes `MarkFSResizeRequired` to set according `mountedPod.fsResizeRequired` to true.
```

Comment:
It seems odd that the dsw populator is checking for resize and then modifying asw.

Would it align better architecturally, if:
- dsw gets updated with pvc.status
- reconciler sees the difference between dsw and asw, triggers resize
- resize operation updates size afterwards

Reply:
It's a copy of the `IsVolumeNotAttachedError` and `IsRemountRequiredError` pattern (see clauses https://github.com/kubernetes/kubernetes/blob/3624c74ce880168eb456f15ef50a92d1cbff8da1/pkg/kubelet/volumemanager/reconciler/reconciler.go#L190)

It is definitely odd that we modify *actual*-sw to say 'resize/remount/attach is *desired*' but practically it doesn't matter because either way the reconciler is looking for a difference between asw and dsw. The `fsResizeRequired` flag is equivalent to having pvc.status in the dsw because the pvc.status vs pv.spec check is done when the pod+volume is added to dsw and if we were to do the check later at the reconciler it would get the same result for `fsResizeRequired` anyway.

Reply:
How difficult is it to move to the other model?  I think we want to avoid overstepping interface boundaries as much as possible.  In the past, there have been tricky race conditions and ordering constraints introduced due to deviation from the intended design, and we've had to add many one-off bandaids to fix the various issues.

Reply:
I think it's doable. But IDK if I can rewrite the KEP in time, or if that's even necessary (can I edit it to match the new implementation after the new implementation is merged?)

Correct me if anything here sounds wrong, just trying to fill in details for myself:

In the current model, `asw.mountedPod.fsResizeRequired` is touched in 3 places.
- dsw-populator writes it T if needed and it already exists in asw. It periodically does `ReprocessPod` at least as frequently as kubelet does `syncPod` so it will keep the field updated.
- reconciler reads it & checks for T/F.
- operationExecutor writes it F on success.

In the new model...
- dsw-populator writes `dsw.volumeTomount.pvcStatusCapacity` always and keeps it updated, similar to above.
- the volume is mounted, the field gets propagated to `asw.mountedPod.pvcstatusCapacity`
- reconciler reads the 2 fields & checks for difference
- operationExecutor writes `asw.mountedPod.pvcstatusCapacity` to the new value on success.

Also, I would want to move remountRequired as well at some later date. nvm about volumeNotAttached, it's the one that makes sense and that remountRequired/fsResizeRequired shouldn't have piggy-backed off of.

Reply:
wait, it's more like dsw.volumeTomount.pvSpecSize and asw.mountedPod.pvcstatusCapacity.

Reply:
Yup that makes sense to me! I think we can actually remove all of the code snippets from the KEP, and just talk about it at a higher level like you just did

Reply:
ok, removed code & wrote this out again. Yeah, don't want to be too prescriptive, details of what goes where can be figured out during actual coding, but the gist as i understand it is to avoid dswp touching asw.